### PR TITLE
Forward headers from Grafana to GraphQL requests

### LINKED ISSUE #7: Headers of Grafana requests aren't forwarded to the HTTP request
Headers of Grafana requests which include Cookie and OAuth identity aren't forwarded to HTTP requests of the data source.

### PR DESCRIPTION
**Pull request by hgiasac**
fix https://github.com/wildmountainfarms/wild-graphql-datasource/issues/7

Forward headers from Grafana requests to upstream GraphQL requests

### CODE REVIEW

File: pkg/plugin/datasource.go [hunk 1]
```diff
@@ -133,6 +134,9 @@ func (d *Datasource) query(ctx context.Context, pCtx backend.PluginContext, quer
 		return nil, err
 	}
 
+	for key, value := range req.Headers {
```
Comment (retrodaredevil) [inline pkg/plugin/datasource.go]: I'll admit I'm unfamiliar with this part of the backend, but I was looking around and I noticed that the function `GetHTTPHeaders()` also exists. I'm wondering if we are supposed to use that instead or what the different would be.

Comment (hgiasac) [reply pkg/plugin/datasource.go]: You're right. The GetHTTPHeaders method filters unused headers ([source](https://github.com/grafana/grafana-plugin-sdk-go/blob/6ae7edaf09e3a007d3738e78b3492a4f6bb515b7/backend/http_headers.go#L66)). We can use that method for the better security.

**Comment by retrodaredevil** [comment]
Can I ask what your use case for this is? I'm also curious if this is standard to pass headers to the server the backend is communicating with. Do other Grafana plugins do this?

My guess is that this is useful so that you can tell which user a particular query has come from? Do you know if these headers can be spoofed by the frontend, or are these headers validated by Grafana before passing them to us?

I ask this because I'm curious if you have any additional knowledge that might help me understand if this is something that many Grafana plugins do (the documentation for this sort of thing is non-existent).

**Comment by retrodaredevil** [comment]
Additionally, I'm wondering if there's a chance that making a change like this could unintentionally pass headers to a GraphQL server that really don't need to be passed (I mean, there are public GraphQL servers out there, and GraphQL serveres that might become unhappy if you pass invalid headers to them). This change seems useful, but I'm wondering if we should have a toggle for it. (I'm not saying to add a toggle yet, just want some discussion around this)

**Comment by hgiasac** [comment]
@retrodaredevil This issue happens in other data sources as well. We need to pass Cookie or OAuth identity for the session auth.

https://github.com/grafana/grafana/issues/60510
https://community.grafana.com/t/elasticsearch-datasource-withcredentials-doesnt-seem-to-send-cookie/5063 

You already knew that Grafana had provided the new HTTP auth UI. Those auth headers are also configurable in the data source config. Therefore that should be fine I guess.
graphql

**Comment by retrodaredevil** [comment]
Oh that makes sense. Have you confirmed that Grafana will automatically filter out headers if we don't have any of those checkboxes checked? We don't have to do the filtering of headers manually, right?

**Comment by hgiasac** [comment]
Yes, as you can see in the [source code](https://github.com/grafana/grafana-plugin-sdk-go/blob/6ae7edaf09e3a007d3738e78b3492a4f6bb515b7/backend/http_headers.go#L66), the `GetHTTPHeaders` function filters the following headers:
- Cookie
- Authorization
- X-Id-Token
- Custom headers with `http_` prefix

```go
func getHTTPHeadersFromStringMap(headers map[string]string) http.Header {
	httpHeaders := http.Header{}

	for k, v := range headers {
		if textproto.CanonicalMIMEHeaderKey(k) == OAuthIdentityTokenHeaderName {
			httpHeaders.Set(k, v)
		}

		if textproto.CanonicalMIMEHeaderKey(k) == OAuthIdentityIDTokenHeaderName {
			httpHeaders.Set(k, v)
		}

		if textproto.CanonicalMIMEHeaderKey(k) == CookiesHeaderName {
			httpHeaders.Set(k, v)
		}

		if strings.HasPrefix(k, httpHeaderPrefix) {
			hKey := strings.TrimPrefix(k, httpHeaderPrefix)
			httpHeaders.Set(hKey, v)
		}
	}

	return httpHeaders
}
```

**Comment by retrodaredevil** [comment]
Yeah I saw that. I'm wondering if the headers present within `req.Headers` will have already been filtered based on whether or not the user has checked those check boxes.

For instance, if a user does not have the "Forward OAuth Identity" checked, do we have to be the ones to not send the header, or will Grafana have already filtered it for us?

**Comment by hgiasac** [comment]
As I read from Grafana's source code, Grafana already filters those headers before sending them to the data source.
Even if there are unexpected headers from Grafana that would be a Grafana's bug, not from us.

**Comment by retrodaredevil** [comment]
Awesome. If you have a source for where it's doing that filtering, I'd appreciate a link so I can look at it. (I'm just curious)

As an additional note, it looks like even with all those checkboxes unchecked, it looks like for requests from a query, these headers get passed each time:
* `X-Datasource-Uid`
* `X-Grafana-Org-Id`
* `X-Panel-Id`
* `X-Dashboard-Uid`

Just wanted to document that those are the headers I observed being passed to the GraphQL server.

Obviously most GraphQL servers aren't going to utilize those additional headers, so it shouldn't break anything. I do wonder if it's a bad practice to pass those headers to all GraphQL servers, but I'm not convinced we need to introduce a toggle right now.

Looks good to me. I think I'll merge this now.

**Comment by hgiasac** [comment]
@retrodaredevil you can look at [plugin integration middleware](https://github.com/grafana/grafana/blob/9125f0df20b80a1053fa9320619bde22d33b8131/pkg/services/pluginsintegration/pluginsintegration.go#L188)

Each middleware applies validation for those auth headers

**Pull request merged**